src/balloonpanel/balloonpanelview.js

```diff
@@ -177,30 +178,37 @@ export default class BalloonPanelView extends View {
 			} )
 		};
 
-		this._smartAttachTo( possiblePanelRects, visibleRect );
+		this._smartAttachTo( possiblePanelRects, limiterVisibleRect, panelRect.width * panelRect.height );
 	}
 
 	/**
-	 * For the given `Array` of possible rects, choses the one which fits the best into
-	 * `visibleViewportRect`, which is when their intersection has the biggest area.
+	 * For the given set of possible rects, chooses the one which fits the best into both - browser viewport and
+	 * `visibleContainerRect`, which is when their intersection has the biggest area. Note that priority is a possible
+	 * highest intersection area with browser viewport.
 	 *
 	 * @private
-	 * @param {Array<{AbsoluteDomRect}>} rects List of positions where balloon can be placed.
-	 * @param {AbsoluteDomRect} visibleViewportRect The absolute rect of the visible part of the browser viewport.
+	 * @param {Object} rects Set of positions where balloon can be placed.
+	 * @param {AbsoluteDomRect} visibleContainerRect The absolute rect of the visible part of container element.
 	 */
-	_smartAttachTo( rects, visibleViewportRect ) {
+	_smartAttachTo( rects, visibleContainerRect, panelSurfaceArea ) {
+		const viewportRect = new AbsoluteDomRect( getViewportRect() );
 		let maxIntersectRectPos;
-		let maxIntersectArea = -1;
+		let maxContainerIntersectArea = -1;
+		let maxViewportIntersectArea = -1;
 
 		// Get best place.
-		for ( let rectPos in rects ) {
-			const intersectArea = rects[ rectPos ].getIntersectArea( visibleViewportRect );
+		Object.keys( rects ).some( ( rectPos ) => {
+			const containerIntersectArea = rects[ rectPos ].getIntersectArea( visibleContainerRect );
+			const viewportIntersectArea = rects[ rectPos ].getIntersectArea( viewportRect );
 
-			if ( intersectArea > maxIntersectArea ) {
+			if ( viewportIntersectArea >= maxViewportIntersectArea && containerIntersectArea > maxContainerIntersectArea ) {
 				maxIntersectRectPos = rectPos;
-				maxIntersectArea = intersectArea;
+				maxContainerIntersectArea = containerIntersectArea;
+				maxViewportIntersectArea = viewportIntersectArea;
 			}
-		}
+
+			return maxContainerIntersectArea === panelSurfaceArea;
```

Comment:
I guess you decided to use `Array.some` and `panelSurfaceArea` to stop checking if there's a rect, which is geometrically fully within the `visibleContainerRect`. But I have 2 doubts:

1. Is it really worth it, I mean the additional parameter to the method, if there are only 4 positions to be checked? It's pure JS, not DOM, so I guess it does not improve the performance that much, really. KISS first, optimize next, that's what I always think about such things.
2. If I got the algorithm right, it breaks immediately when such a position in `rects` is found that `maxContainerIntersectArea === panelSurfaceArea`. What about the viewport intersection in such case? 

  For instance, it may happen that the very first position in `rects` matches `maxContainerIntersectArea === panelSurfaceArea` but it is hardly visible in the viewport :P So the algorithm stops and is satisfied with the position, which is the worst for the UX. There could be many `rects` that meet `maxContainerIntersectArea === panelSurfaceArea` but only the last one is actually completely within the viewport. 

  I'm afraid that we need to sort `rects` by `viewportIntersectArea` first, and then secondly by `containerIntersectArea`. To do this, the algorithm must analyze all the `rects`, so no `some()`. 

  The algorithm should search among the `rects` which are completely in the viewport first for such with the biggest `containerIntersectArea`. If there are no `rects` completely within the viewport, then the `containerIntersectArea` should be ignored and the one which fits the most in the viewport is the best. I guess ;-)